### ISSUE TITLE
Fits file doesn't load when header doesn't adhere to fits standard

### ISSUE THREAD

**Issue by lh-astro** [CARTAvis/carta-backend#596]
Copy from @veggiesaurus:

There are a couple of problems with the FITS header.
(for ref: table 7 of the FITS standard: https://fits.gsfc.nasa.gov/standard40/fits_standard40aa-le.pdf)
1. Images need to have headers in the right order (NAXIS, then NAXIS1, ...NAXIS). This has (a) a gap between NAXIS values and (b) additional NAXIS headers (it has up to 4)
2. Then there also appears to be a CASA bug, where including additional CDELT, CRPIX, etc prevent the file from being read. Easiest thing to do is to trim the extra axis 3 and 4 headers, or pad them:
```
SIMPLE  =                    T                                                  
BITPIX  =                  -64                                                  
NAXIS   =                    4                                                  
NAXIS1  =                 9203                                                  
NAXIS2  =                 4706                                                  
NAXIS3  =                    1                                                  
NAXIS4  =                    1                                                  
EXTEND  =                    T / FITS dataset may contain extensions            
COMMENT   FITS (Flexible Image Transport System) format is defined in 'Astronomy
COMMENT   and Astrophysics', volume 376, page 359; bibcode: 2001A&A...376..359H 
CTYPE1  = 'RA---TAN'                                                            
CTYPE2  = 'DEC--TAN'                                                            
CTYPE3  = 'FREQ'                                                                
CTYPE4  = 'STOKES'                                                              
EQUINOX =                 2000                                                  
CRVAL1  =         35.152236620                                                  
CRVAL2  =         -4.833701903                                                  
CRVAL3  =          1.000000000                                                  
CRVAL4  = 1283895507.812500000                                                  
CRPIX1  =               4601.5                                                  
CRPIX2  =                2352.                                                  
CRPIX3  =          1.000000000                                                  
CRPIX4  =          1.000000000                                                  
CDELT1  =         -0.000305556                                                  
CDELT2  =          0.000305556                                                  
CDELT3  =          1.000000000                                                  
CDELT4  =          1.000000000                                                  
CROTA2  =        359.999992933
```
and
```
SIMPLE  =                    T                                                  
BITPIX  =                  -64                                                  
NAXIS   =                    2                                                  
NAXIS1  =                 9203                                                  
NAXIS2  =                 4706                                                  
EXTEND  =                    T / FITS dataset may contain extensions            
COMMENT   FITS (Flexible Image Transport System) format is defined in 'Astronomy
COMMENT   and Astrophysics', volume 376, page 359; bibcode: 2001A&A...376..359H 
CTYPE1  = 'RA---TAN'                                                            
CTYPE2  = 'DEC--TAN'                                                            
EQUINOX =                 2000                                                  
CRVAL1  =         35.152236620                                                  
CRVAL2  =         -4.833701903                                                  
CRPIX1  =               4601.5                                                  
CRPIX2  =                2352.                                                  
CDELT1  =         -0.000305556                                                  
CDELT2  =          0.000305556                                                  
CROTA2  =        359.999992933
```

both work
(1) is a FITS standard, (2) is a CASA bug. We could try submit a PR to casacore to fix (2), I think. But (1) is probably best adhered to

**Confirm:** Re-formatting the header as a work-around solves the issue.

**Comment by veggiesaurus** [comment]
@Jordatious had a similar issue a while back, I think. Is there a specific pipeline producing these images by any chance? 

I do think we should try to fix the behaviour when there are extra _CRVAL_ or _CRPIX_ headers. As far as I know, this isn't forbidden by the FITS standard

**Comment by pford** [comment]
@veggiesaurus do you have a test FITS image for this issue?

**Comment by Jordatious** [comment]
> @Jordatious had a similar issue a while back, I think. Is there a specific pipeline producing these images by any chance?

Not that I know. I've seen basically identical headers from Ian Heywood after mosaicking pointings together. I could ask him?

**Comment by veggiesaurus** [comment]
> > @Jordatious had a similar issue a while back, I think. Is there a specific pipeline producing these images by any chance?
> 
> Not that I know. I've seen basically identical headers from Ian Heywood after mosaicking pointings together. I could ask him?

Please do. @arusstaylor also complained about an image not loading, which turned out to have the exact same type of invalid header entries.

**Comment by veggiesaurus** [comment]
> @veggiesaurus do you have a test FITS image for this issue?

@pford have sent you an image on slack

**Comment by Jordatious** [comment]
> Please do. @arusstaylor also complained about an image not loading, which turned out to have the exact same type of invalid header entries.
 
I've asked. I spent way too long trying to fix the same kind of FITS header for a particular image last week…

**Comment by Jordatious** [comment]
[Montage](http://montage.ipac.caltech.edu/)

**Comment by Jordatious** [comment]
Sweet! Is this generally solved, or just for FITS headers like the above?

**Comment by veggiesaurus** [comment]
> Sweet! Is this generally solved, or just for FITS headers like the above?

@kswang1029 can you confirm?

**Comment by veggiesaurus** [comment]
Re-opening until @kswang1029 has done some more testing with unordered FITS headers (specifically `NAXIS` headers)

**Comment by kswang1029** [comment]
Here are some test results with `origin/pam/596_incomplete_fits_header 4fcb6fa`

1. If the four axes are ordered as RA-Dec-Stokes-Freq in the header, it works properly. 
2. If the four axes are ordered as RA-Dec-Freq-Stokes in the header but with CTYPE3 and CTYPE4 missing, it works properly. The backend treats the 3rd axis as channel and the 4th as Stokes in the _animator_. However, frequency/velocity info and Stokes info are missing in the cursor info bar. This is acceptable I think.
3. If the four axes are ordered as RA-Dec-Sponge-Bob in the header, it works with the "spectral" axis displayed as "Sponge" in the cursor info bar. The Sponge axis and the Bob axis are recognized as channel and Stokes in the animator, respectively.
4. With a header like 
<img width="468" alt="Screen Shot 2021-01-26 at 2 09 29 PM" src="https://user-images.githubusercontent.com/20819712/105807055-27cd2180-5fe0-11eb-80c5-14dd271c05a8.png">
where C****4 as "Stokes" is listed before C****3 as "Frequency", it works but after casacore, the 3rd axis is Stokes and the 4th is frequency. Animator and cursor info bar work fine.

5. With a header like 
<img width="460" alt="Screen Shot 2021-01-26 at 2 12 51 PM" src="https://user-images.githubusercontent.com/20819712/105807258-9dd18880-5fe0-11eb-9747-a654891eb2ba.png">
where 4 axes are claimed but with missing C****3 and C****4, the image loads fine (only the first channel). The animator does not show channel nor Stokes. This is acceptable I think. With v1.4, this kind of image cannot be loaded, not even getting file info.

6. I tried to generate a 4D image with NAXIS3 and NAXIS4 not listed in the header but astropy is too smart not allowing me to get header I intended to have. The final image shows NAXIS3 and NAXIS4 with correct numbers. 

Overall, I think the new approach works fine. 

@Jordatious could you give it a try or if possible pass me those problematic images to test with?

**Comment by Jordatious** [comment]
Awesome! I have one HI cube that doesn't work. Here's the header:

```
SIMPLE  =                    T                                                  
BITPIX  =                  -32                                                  
NAXIS   =                    3                                                  
NAXIS1  =                  370                                                  
NAXIS2  =                  514                                                  
NAXIS3  =                   74                                                  
CTYPE1  = 'GLON-FLT'                                                            
CRVAL1  =  1.3470000000000E+02                                                  
CRPIX1  =  1.8775000000000E+02                                                  
CDELT1  = -1.6666665425876E-02                                                  
CTYPE2  = 'GLAT-FLT'                                                            
CRVAL2  =  3.6000000000000E+00                                                  
CRPIX2  =  2.1325000000000E+02                                                  
CDELT2  =  1.6666665425876E-02                                                  
CTYPE3  = 'VELO-LSR'                                                            
CRVAL3  = -5.0000000000000E+04                                                  
CRPIX3  =  3.7000000000000E+01                                                  
CDELT3  = -1.6490000486374E+03                                                  
BUNIT   = 'K       '                                                            
DATAMIN = -1.0926479492188E+03                                                  
DATAMAX =  1.5325633239746E+02                                                  
HISTORY kview: Module version 1.1.22  Karma v1.7.7  compiled with v1.7.7      
```

**Comment by veggiesaurus** [comment]
> Awesome! I have one HI cube that doesn't work. Here's the header:
> 
> ```
> SIMPLE  =                    T                                                  
> BITPIX  =                  -32                                                  
> NAXIS   =                    3                                                  
> NAXIS1  =                  370                                                  
> NAXIS2  =                  514                                                  
> NAXIS3  =                   74                                                  
> CTYPE1  = 'GLON-FLT'                                                            
> CRVAL1  =  1.3470000000000E+02                                                  
> CRPIX1  =  1.8775000000000E+02                                                  
> CDELT1  = -1.6666665425876E-02                                                  
> CTYPE2  = 'GLAT-FLT'                                                            
> CRVAL2  =  3.6000000000000E+00                                                  
> CRPIX2  =  2.1325000000000E+02                                                  
> CDELT2  =  1.6666665425876E-02                                                  
> CTYPE3  = 'VELO-LSR'                                                            
> CRVAL3  = -5.0000000000000E+04                                                  
> CRPIX3  =  3.7000000000000E+01                                                  
> CDELT3  = -1.6490000486374E+03                                                  
> BUNIT   = 'K       '                                                            
> DATAMIN = -1.0926479492188E+03                                                  
> DATAMAX =  1.5325633239746E+02                                                  
> HISTORY kview: Module version 1.1.22  Karma v1.7.7  compiled with v1.7.7      
> ```

I'm not sure if this is the same issue though... This header looks 💯 to me, unless I'm missing something.

**Comment by Jordatious** [comment]
The previous Montage image I had is able to open just fine now. 

One issue with the other header could be CARTA simply doesn't associate with Karma 😂

**Comment by kswang1029** [comment]
> Awesome! I have one HI cube that doesn't work. Here's the header:
> 
> ```
> SIMPLE  =                    T                                                  
> BITPIX  =                  -32                                                  
> NAXIS   =                    3                                                  
> NAXIS1  =                  370                                                  
> NAXIS2  =                  514                                                  
> NAXIS3  =                   74                                                  
> CTYPE1  = 'GLON-FLT'                                                            
> CRVAL1  =  1.3470000000000E+02                                                  
> CRPIX1  =  1.8775000000000E+02                                                  
> CDELT1  = -1.6666665425876E-02                                                  
> CTYPE2  = 'GLAT-FLT'                                                            
> CRVAL2  =  3.6000000000000E+00                                                  
> CRPIX2  =  2.1325000000000E+02                                                  
> CDELT2  =  1.6666665425876E-02                                                  
> CTYPE3  = 'VELO-LSR'                                                            
> CRVAL3  = -5.0000000000000E+04                                                  
> CRPIX3  =  3.7000000000000E+01                                                  
> CDELT3  = -1.6490000486374E+03                                                  
> BUNIT   = 'K       '                                                            
> DATAMIN = -1.0926479492188E+03                                                  
> DATAMAX =  1.5325633239746E+02                                                  
> HISTORY kview: Module version 1.1.22  Karma v1.7.7  compiled with v1.7.7      
> ```

I can try to mock up an image cube based on this header and try.

**Comment by kswang1029** [comment]
I mocked one up but it does not load. 
<img width="522" alt="Screen Shot 2021-01-26 at 6 50 20 PM" src="https://user-images.githubusercontent.com/20819712/105835989-8ce83d80-6007-11eb-93a1-375efa6f017a.png">
<img width="535" alt="Screen Shot 2021-01-26 at 6 49 53 PM" src="https://user-images.githubusercontent.com/20819712/105836042-9ec9e080-6007-11eb-8e81-d239d6b1d9c6.png">
@Jordatious do you see the same error message from CARTA too?

The backend log says (nothing critical I think)

> 2021-01-26 10:44:22     INFO    FITSCoordinateUtil::fromFITSHeader      Header
2021-01-26 10:44:22     INFO    FITSCoordinateUtil::fromFITSHeader+     CTYPE1  = 'GLON-FLT'
2021-01-26 10:44:22     INFO    FITSCoordinateUtil::fromFITSHeader+     was interpreted as
2021-01-26 10:44:22     INFO    FITSCoordinateUtil::fromFITSHeader+     CTYPE1  = 'GLON-CAR'                                                            0¥¹^G^A
2021-01-26 10:44:22     INFO    FITSCoordinateUtil::fromFITSHeader      Header
2021-01-26 10:44:22     INFO    FITSCoordinateUtil::fromFITSHeader+     CTYPE2  = 'GLAT-FLT'
2021-01-26 10:44:22     INFO    FITSCoordinateUtil::fromFITSHeader+     was interpreted as
2021-01-26 10:44:22     INFO    FITSCoordinateUtil::fromFITSHeader+     CTYPE2  = 'GLAT-CAR'                                                            0¥¹^G^A
2021-01-26 10:44:22     WARN    FITSCoordinateUtil::fromFITSHeader      Zero or no rest frequency provided for velocity axis.
2021-01-26 10:49:19     INFO    FITSCoordinateUtil::fromFITSHeader      Header
2021-01-26 10:49:19     INFO    FITSCoordinateUtil::fromFITSHeader+     CTYPE1  = 'GLON-FLT'
2021-01-26 10:49:19     INFO    FITSCoordinateUtil::fromFITSHeader+     was interpreted as
2021-01-26 10:49:19     INFO    FITSCoordinateUtil::fromFITSHeader+     CTYPE1  = 'GLON-CAR'                                                            0¥¹^G^A
2021-01-26 10:49:19     INFO    FITSCoordinateUtil::fromFITSHeader      Header
2021-01-26 10:49:19     INFO    FITSCoordinateUtil::fromFITSHeader+     CTYPE2  = 'GLAT-FLT'
2021-01-26 10:49:19     INFO    FITSCoordinateUtil::fromFITSHeader+     was interpreted as
2021-01-26 10:49:19     INFO    FITSCoordinateUtil::fromFITSHeader+     CTYPE2  = 'GLAT-CAR'                                                            0¥¹^G^A
2021-01-26 10:49:19     WARN    FITSCoordinateUtil::fromFITSHeader      Zero or no rest frequency provided for velocity axis.



Note that this case missed units for each axis as well as other headers.

> #hdr.set('BSCALE', 1.000000000000E+00, 'PHYSICAL = PIXEL*BSCALE + BZERO')
#hdr.set('BZERO', 0.000000000000E+00)
#hdr.set('BMAJ', 4.180019088465E-05)
#hdr.set('BMIN', 3.012913555345E-05)
#hdr.set('BPA', 1.151531397762E+01)
#hdr.set('BTYPE', 'intensity')
hdr.set('BUNIT', 'K       ')
#hdr.set('RADESYS', 'ICRS')
#hdr.set('LONPOLE', 1.800000000000E+02)
#hdr.set('LATPOLE', 1.798947222223E+01)
hdr.set('NAXIS', 3)
hdr.set('NAXIS1', n_pix_x)
hdr.set('NAXIS2', n_pix_y)
hdr.set('NAXIS3', n_chan)
hdr.set('CTYPE1', 'GLON-FLT')
hdr.set('CRVAL1', 1.3470000000000E+02)
hdr.set('CDELT1', -1.6666665425876E-02)
hdr.set('CRPIX1', 1.8775000000000E+02)
#hdr.set('CUNIT1', 'deg')
hdr.set('CTYPE2', 'GLAT-FLT')
hdr.set('CRVAL2', 3.6000000000000E+00)
hdr.set('CDELT2', 1.6666665425876E-02)
hdr.set('CRPIX2', 2.1325000000000E+02)
#hdr.set('CUNIT2', 'deg')
hdr.set('CTYPE3', 'VELO-LSR')
hdr.set('CRVAL3', -5.0000000000000E+04)
hdr.set('CDELT3', -1.6490000486374E+03)
hdr.set('CRPIX3', 3.7000000000000E+01)
#hdr.set('CUNIT3', 'Hz')
hdr.set('DATAMIN', -1.0926479492188E+03)
hdr.set('DATAMAX', 1.5325633239746E+02)
hdr.set('HISTORY', 'kview: Module version 1.1.22  Karma v1.7.7  compiled with v1.7.7')
#hdr.set('RESTFREQ', 1.000000000000E+11)
#hdr.set('SPECSYS', 'LSRK')

**Comment by Jordatious** [comment]
>@Jordatious do you see the same error message from CARTA too?

Yep

**Comment by kswang1029** [comment]
Ok we have an issue for that error message already https://github.com/CARTAvis/carta-backend/issues/373

**Comment by Jordatious** [comment]
Sorry @kswang1029, I must admit I didn't really read the error msg, since I've seen it a bunch of times before. I figured CARTA was just saying it couldn't open it and that this was one and the same issue. I'll keep an eye on the other issue.

**Comment by pford** [comment]
@kswang1029 can this issue be closed or does it need further testing?

**Comment by kswang1029** [comment]
@Jordatious are you happy with the new fix here? Any other cases that carta should be able to load but cannot?

**Comment by Jordatious** [comment]
My only outstanding case is the one you suggested is addressed by #373. I'm not 100% about the details here, but it seems fine to close this now and follow up the other issue there.

**Comment by kswang1029** [comment]
@pford @Jordatious ok let's close this issue.
